docs/generate-image-and-sdk.1
'\" t
.\"     Title: generate-image-and-sdk
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.10
.\"      Date: 2022-07-09
.\"    Manual: Generate-image-and-sdk Manual
.\"    Source: 
.\"  Language: English
.\"
.TH "GENERATE\-IMAGE\-AND\-SDK" "1" "2022-07-09" "" "Generate\-image\-and\-sdk Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
generate\-image\-and\-sdk \- Generate and build a python SDK using an embedded shell script. Pass parameters down to the shell script using the parameters below. This command may require additional tools such as graalvm, maven and a local compiler such as gcc to run correctly.
.SH "SYNOPSIS"
.sp
\fBgenerate\-image\-and\-sdk\fP [\fB\-\-buildSharedLibrary\fP] [\fB\-\-enableJetsonNano\fP] [\fB\-\-noGc\fP]
[\fB\-\-binaryExtension\fP=\fI<binaryExtension>\fP]
[\fB\-\-buildPlatform\fP=\fI<buildPlatform>\fP]
[\fB\-\-bundleOutputPath\fP=\fI<bundleOutputPath>\fP]
[\fB\-\-imageName\fP=\fI<imageName>\fP] [\fB\-\-includePath\fP=\fI<includePath>\fP]
[\fB\-\-kompileCPath\fP=\fI<kompileCPath>\fP]
[\fB\-\-kompilePrefix\fP=\fI<kompilePrefix>\fP]
[\fB\-\-kompilePythonPath\fP=\fI<kompilePythonPath>\fP]
[\fB\-\-libOutputPath\fP=\fI<libOutputPath>\fP]
[\fB\-\-mainClass\fP=\fI<mainClass>\fP] [\fB\-\-mavenHome\fP=\fI<mavenHome>\fP]
[\fB\-\-maxRamMegs\fP=\fI<maxRamMegs>\fP] [\fB\-\-minRamMegs\fP=\fI<minRamMegs>\fP]
[\fB\-\-nativeImageFilesPath\fP=\fI<nativeImageFilesPath>\fP]
[\fB\-\-nd4jBackend\fP=\fI<nd4jBackend>\fP]
[\fB\-\-nd4jClassifier\fP=\fI<nd4jClassifier>\fP]
[\fB\-\-pipelineFile\fP=\fI<pipelineFile>\fP]
[\fB\-\-pomGenerateOutputPath\fP=\fI<pomGenerateOutputPath>\fP]
[\fB\-\-protocol\fP=\fI<protocol>\fP]
[\fB\-\-pythonExecutable\fP=\fI<pythonExecutable>\fP]
.SH "DESCRIPTION"
.sp
Generate and build a python SDK using an embedded shell script. Pass parameters down to the shell script using the parameters below. This command may require additional tools such as graalvm, maven and a local compiler such as gcc to run correctly.
.SH "OPTIONS"
.sp
\fB\-\-binaryExtension\fP=\fI<binaryExtension>\fP
.RS 4
The platform to build for, usually a javacpp.platform value such as linux\-x86_64
.RE
.sp
\fB\-\-buildPlatform\fP=\fI<buildPlatform>\fP
.RS 4
The platform to build for, usually a javacpp.platform value such as linux\-x86_64
.RE
.sp
\fB\-\-buildSharedLibrary\fP
.RS 4
Whether to build a shared library or not, defaults to true
.RE
.sp
\fB\-\-bundleOutputPath\fP=\fI<bundleOutputPath>\fP
.RS 4
Path to output file of complete bundle
.RE
.sp
\fB\-\-enableJetsonNano\fP
.RS 4
Whether to use jetson nano dependencies or not
.RE
.sp
\fB\-\-imageName\fP=\fI<imageName>\fP
.RS 4
Name of image output file
.RE
.sp
\fB\-\-includePath\fP=\fI<includePath>\fP
.RS 4
Location of include path for compilation/linking
.RE
.sp
\fB\-\-kompileCPath\fP=\fI<kompileCPath>\fP
.RS 4
Path to kompile c library
.RE
.sp
\fB\-\-kompilePrefix\fP=\fI<kompilePrefix>\fP
.RS 4
The kompile prefix where the relevant kompile source code is for compilation.
.RE
.sp
\fB\-\-kompilePythonPath\fP=\fI<kompilePythonPath>\fP
.RS 4
Path to kompile python sdk
.RE
.sp
\fB\-\-libOutputPath\fP=\fI<libOutputPath>\fP
.RS 4
Location of where to put c library after compilation
.RE
.sp
\fB\-\-mainClass\fP=\fI<mainClass>\fP
.RS 4
The entry point to use in the image
.RE
.sp
\fB\-\-mavenHome\fP=\fI<mavenHome>\fP
.RS 4
The maven home location for compiling native image
.RE
.sp
\fB\-\-maxRamMegs\fP=\fI<maxRamMegs>\fP
.RS 4
The maximum memory usage for the image
.RE
.sp
\fB\-\-minRamMegs\fP=\fI<minRamMegs>\fP
.RS 4
The minimum memory usage for the image
.RE
.sp
\fB\-\-nativeImageFilesPath\fP=\fI<nativeImageFilesPath>\fP
.RS 4
The path to the files for building an image
.RE
.sp
\fB\-\-nd4jBackend\fP=\fI<nd4jBackend>\fP
.RS 4
The nd4j backend to use in the image
.RE
.sp
\fB\-\-nd4jClassifier\fP=\fI<nd4jClassifier>\fP
.RS 4
The nd4j classifier to use
.RE
.sp
\fB\-\-noGc\fP
.RS 4
Whether to use gc in the image or not
.RE
.sp
\fB\-\-pipelineFile\fP=\fI<pipelineFile>\fP
.RS 4
Whether to use a pipeline file or not
.RE
.sp
\fB\-\-pomGenerateOutputPath\fP=\fI<pomGenerateOutputPath>\fP
.RS 4
Output path of the generated pom.xml for compiling native image
.RE
.sp
\fB\-\-protocol\fP=\fI<protocol>\fP
.RS 4
Protocol to use for serving, http or grpc are valid
.RE
.sp
\fB\-\-pythonExecutable\fP=\fI<pythonExecutable>\fP
.RS 4
The executable to use with python. Defaults to the python found on the path. Otherwise will use the built in python installed with ./kompile install python
.RE

src/main/resources/generate-image-and-sdk.sh
#!/bin/bash

#
# Copyright (c) 2022 Konduit K.K.
#
#     This program and the accompanying materials are made available under the
#     terms of the Apache License, Version 2.0 which is available at
#     https://www.apache.org/licenses/LICENSE-2.0.
#
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#     WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#     License for the specific language governing permissions and limitations
#     under the License.
#
#     SPDX-License-Identifier: Apache-2.0
#

set -eu


# Use > 1 to consume two arguments per pass in the loop (e.g. each
# argument has a corresponding value to go with it).
# Use > 0 to consume one or more arguments per pass in the loop (e.g.
# some arguments don't have a corresponding value to go with it such
# as in the --default example).
# note: if this is set to > 0 the /etc/hosts part is not recognized ( may be a bug )
PIPELINE_FILE=
KOMPILE_PYTHON_PATH="../kompile-python"
KOMPILE_C_PATH="../kompile-c-library"
IMAGE_NAME="konduit-serving"
# http,grpc
PROTOCOL=
POM_GENERATE_OUTPUT_PATH="pom2.xml"
# Where to put the common libraries output from native image, c libraries
LIB_OUTPUT_PATH="./lib"
INCLUDE_PATH="./include"
# Where to output final  bundle with python sdk, compiled binaries, install script, pipeline file/model assets
BUNDLE_OUTPUT_PATH="${IMAGE_NAME}-pipeline-bundle.zip"
MAVEN_HOME=
BUILD_PLATFORM=
BINARY_EXTENSION=
ND4J_BACKEND="nd4j-native"
ND4J_CLASSIFIER=
ND4J_HELPER=
ND4J_DATATYPES=
ND4J_OPERATIONS=
ENABLE_JETSON_NANO="false"
BUILD_SHARED_LIBRARY="true"
MAIN_CLASS=
MIN_RAM_MEGS=2000
MAX_RAM_MEGS=2000
NO_GC="false"
NATIVE_IMAGE_FILE_PATH=
KOMPILE_PREFIX="./"
PYTHON_EXEC="python"
BUILD_CUDA_BACKEND="false"
BUILD_CPU_BACKEND="false"
IS_SERVER="false"


while [[ $# -gt 0 ]]
do
key="$1"
value="${2:-}"
case $key in
    -p|--pipeline-file)
    PIPELINE_FILE="$value"
    shift # past argument
    ;;
     -dt|--nd4j-datatypes)
      ND4J_DATATYPES="$value"
      shift # past argument
      ;;
     -s|--server)
        IS_SERVER="$value"
        shift # past argument
      ;;
      -op|--nd4j-operations)
        ND4J_OPERATIONS="$value"
        shift # past argument
      ;;
    -py|-python-sdk|--python-sdk)
    KOMPILE_PYTHON_PATH="$value"
    shift # past argument
    ;;
    -c|--c-library)
    KOMPILE_C_PATH="$value"
    shift # past argument
    ;;
    -s|--is-server)
      IS_SERVER="$value"
      shift # past argument
      ;;
    -pe|--python-exec)
      PYTHON_EXEC="$value"
      shift # past argument
      ;;
    -i|--image-name)
    IMAGE_NAME="$value"
    shift # past argument
    ;;
    -pl|--protocol)
    PROTOCOL="$value"
    shift # past argument
    ;;
    -pom|--pom-path)
    POM_GENERATE_OUTPUT_PATH="$value"
    shift # past argument
    ;;
    -lp|--lib-path)
    LIB_OUTPUT_PATH="$value"
    shift # past argument
    ;;
    -mh|--maven-home)
    MAVEN_HOME="$value"
    shift # past argument
    ;;
     -nb|--nd4j-backend)
    ND4J_BACKEND="$value"
    shift # past argument
    ;;
     -nc|--nd4j-classifier)
    ND4J_CLASSIFIER="$value"
    shift # past argument
    ;;
    -nh|--nd4j-helper)
      ND4J_HELPER="$value"
      shift # past argument
      ;;
    -en|--enable-jetson-nano)
    ENABLE_JETSON_NANO="$value"
    shift # past argument
    ;;
    -bs|--build-shared)
    BUILD_SHARED_LIBRARY="$value"
    shift # past argument
    ;;
     -mc|--main-class)
    MAIN_CLASS="$value"
    shift # past argument
    ;;
    -mr|--min-ram)
    MIN_RAM_MEGS="$value"
    shift # past argument
    ;;
    -mar|--max-ram)
    MAX_RAM_MEGS="$value"
    shift # past argument
    ;;
    -nifp|--native-image-file-path)
    NATIVE_IMAGE_FILE_PATH="$value"
    shift # past argument
    ;;
    -kp|--kompile-prefix)
      KOMPILE_PREFIX="$value"
      shift # past argument
      ;;
   -ngc|--no-garbage-collection)
    NO_GC="$value"
    shift # past argument
    ;;
    *)
            # unknown option
    ;;
esac
if [[ $# -gt 0 ]]; then
    shift # past argument or value
fi
done

function set_platform () {
  OS=
  if [[ "$OSTYPE" == "linux-gnu"* ]]; then
        OS="linux"
elif [[ "$OSTYPE" == "darwin"* ]]; then
        OS="mac-osx"
elif [[ "$OSTYPE" == "cygwin" ]]; then
       OS="windows"
elif [[ "$OSTYPE" == "msys" ]]; then
       OS="windows"
elif [[ "$OSTYPE" == "win32" ]]; then
        OS="windows"
elif [[ "$OSTYPE" == "freebsd"* ]]; then
        OS="linux"
else
        OS="linux"
fi

PLATFORM="$(lscpu | grep Architecture | tr -d ':'  | sed 's/Architecture//' | xargs echo -n)"
if [ "${PLATFORM}" == "aarch64" ];then
      PLATFORM="arm64"
fi
BUILD_PLATFORM="${OS}-${PLATFORM}"
}





function set_binary_extension {
if [[ "$OSTYPE" == "linux-gnu"* ]]; then
        BINARY_EXTENSION="so"
elif [[ "$OSTYPE" == "darwin"* ]]; then
        BINARY_EXTENSION="dylib"
elif [[ "$OSTYPE" == "cygwin" ]]; then
       BINARY_EXTENSION="dll"
elif [[ "$OSTYPE" == "msys" ]]; then
       BINARY_EXTENSION="dll"
elif [[ "$OSTYPE" == "win32" ]]; then
        BINARY_EXTENSION="dll"
elif [[ "$OSTYPE" == "freebsd"* ]]; then
        BINARY_EXTENSION="so"
else
         BINARY_EXTENSION="so"
fi

}

#Set a platform default if one isn't found
if [ -z "${BUILD_PLATFORM}" ]; then
  set_platform
fi


if [ -z "${BUILD_PLATFORM}" ]; then
  ND4J_CLASSIFIER="${BUILD_PLATFORM}"
fi

set_binary_extension

# Set this for cmake environment so it automatically detects needed variables
export LIB_OUTPUT_PATH
export INCLUDE_PATH


echo "PIPELINE_FILE ${PIPELINE_FILE}"
echo "KOMPILE_PYTHON_PATH= ${KOMPILE_PYTHON_PATH}"
echo "KOMPILE_C_PATH ${KOMPILE_C_PATH}"
echo "IMAGE_NAME ${IMAGE_NAME}"
echo "PROTOCOL ${PROTOCOL}"
echo "POM_GENERATE_OUTPUT_PATH ${POM_GENERATE_OUTPUT_PATH}"
echo "LIB_OUTPUT_PATH ${LIB_OUTPUT_PATH}"
echo "INCLUDE_PATH ${INCLUDE_PATH}"
echo "BUNDLE_OUTPUT_PATH ${BUNDLE_OUTPUT_PATH}"
echo "MAVEN_HOME ${MAVEN_HOME}"
echo "BUILD_PLATFORM ${BUILD_PLATFORM}"
echo "BINARY_EXTENSION ${BINARY_EXTENSION}"
echo "ND4J_BACKEND ${ND4J_BACKEND}"
echo "ND4J_CLASSIFIER ${ND4J_CLASSIFIER}"
echo "ENABLE_JETSON_NANO ${ENABLE_JETSON_NANO}"
echo "BUILD_SHARED_LIBRARY ${BUILD_SHARED_LIBRARY}"
echo "MAIN_CLASS ${MAIN_CLASS}"
echo "MIN_RAM_MEGS ${MIN_RAM_MEGS}"
echo "MAX_RAM_MEGS ${MAX_RAM_MEGS}"
echo "NO_GC ${NO_GC}"
echo "NATIVE_IMAGE_FILE_PATH ${NATIVE_IMAGE_FILE_PATH}"
echo "IS_SERVER ${IS_SERVER}"

if [ "${ND4J_BACKEND}"  = "nd4j-native" ]; then
      BUILD_CPU_BACKEND="true"
    else
        BUILD_CPU_BACKEND="false"
fi

if [[ "${ND4J_BACKEND}" == *"nd4j-cuda"* ]]; then
      BUILD_CUDA_BACKEND="true"
    else
        BUILD_CUDA_BACKEND="false"
fi

if test -f "$PIPELINE_FILE"; then
    ./kompile build clone-build \
              --dl4jDirectory=${KOMPILE_PREFIX}/deeplearning4j \
              --konduitServingDirectory=${KOMPILE_PREFIX}/konduit-serving \
              --buildDl4j \
              --buildKonduitServing \
              --libnd4jClassifier="${ND4J_CLASSIFIER}" \
              --buildCpuBackend="${BUILD_CPU_BACKEND}" \
               --buildCudaBackend="${BUILD_CUDA_BACKEND}" \
               --libnd4jHelper="${ND4J_HELPER}" \
               --libnd4jOperations="${ND4J_OPERATIONS}" \
               --libnd4jDataTypes="${ND4J_DATATYPES}"
    echo "Processing pipeline file $PIPELINE_FILE"
    echo "Outputting pom file for build to ${POM_GENERATE_OUTPUT_PATH}"
    POM_GENERATE_COMMAND="$(./kompile build  pipeline-command-generate  --server=${IS_SERVER} --nd4jBackend=${ND4J_BACKEND} --nd4jBackendClassifier=${ND4J_CLASSIFIER} --mainClass=${MAIN_CLASS}   --pipelineFile=${PIPELINE_FILE}  --numpySharedLibrary=${BUILD_SHARED_LIBRARY}  --imageName=${IMAGE_NAME}  --outputFile=${POM_GENERATE_OUTPUT_PATH})"
    echo "Command pom generate command was ${POM_GENERATE_COMMAND}"
    eval "./kompile ${POM_GENERATE_COMMAND}"
    BUILD_DIR="$(pwd)"
    export NATIVE_LIB_DIR="${BUILD_DIR}/${IMAGE_NAME}/target"
    ./kompile build native-image-generate  \
                --server="${IS_SERVER}" \
                --nativeImageFilesPath="${NATIVE_IMAGE_FILE_PATH}" \
                --imageName="${IMAGE_NAME}" \
                --outputFile="${POM_GENERATE_OUTPUT_PATH}" \
                --pipelinePath="${PIPELINE_FILE}" \
                --mavenHome="${MAVEN_HOME}" \
                --numpySharedLibrary="${BUILD_SHARED_LIBRARY}" \
                --javacppPlatform="${BUILD_PLATFORM}" \
                --mainClass="${MAIN_CLASS}"
    cd "${NATIVE_LIB_DIR}"

    if [ "${IS_SERVER}" == "false" ]; then
         ln -s ./"${IMAGE_NAME}.so" "lib${IMAGE_NAME}.so"
         cd -
         echo "Creating library directory ${LIB_OUTPUT_PATH} and include directory ${INCLUDE_PATH} if not exists"

         mkdir -p "${LIB_OUTPUT_PATH}"
         cd "${LIB_OUTPUT_PATH}"
         # Resolve absolute path in case relative path is specified
         REAL_LIB_PATH="$(pwd)"
         echo "Set library path to ${REAL_LIB_PATH}"
         cd "${BUILD_DIR}"
         mkdir -p "${INCLUDE_PATH}"
         cd "${INCLUDE_PATH}"
         # Capture absolute path of include directory as well in case relative path is specified
         REAL_INCLUDE_PATH="$(pwd)"
         echo "Set real include path to ${REAL_INCLUDE_PATH}"
         cd "${BUILD_DIR}"
         cp "${KOMPILE_PREFIX}/${IMAGE_NAME}/target/"*.h "${INCLUDE_PATH}"
         cp "${KOMPILE_PREFIX}/src/main/resources/numpy_struct.h" "${INCLUDE_PATH}"
         cp "${KOMPILE_PREFIX}/${IMAGE_NAME}/target/"*.${BINARY_EXTENSION} "${LIB_OUTPUT_PATH}"
          if  test -f './kompile-c' ; then
              rm -rf './kompile-c'
         fi
         # Sometimes CMakeCache.txt maybe present. Remove it before copying to ensure a build proceeds.
         if  test -f "${KOMPILE_PREFIX}/${KOMPILE_C_PATH}/CMakeCache.txt" ; then
              rm -rf "${KOMPILE_PREFIX}/${KOMPILE_C_PATH}/CMakeCache.txt"
         fi
         cp -rf "${KOMPILE_PREFIX}/${KOMPILE_C_PATH}" ./kompile-c

         cd ${KOMPILE_PREFIX}./kompile-c
         cmake .
         make
         # Note we don't quote here so it resolves the binary extension properly
         cp *."${BINARY_EXTENSION}" "${REAL_LIB_PATH}"
         export LIB_OUTPUT_PATH="${REAL_LIB_PATH}"
         cd ..
         # Ensure link path is set for compiling the right python libraries
         export LD_LIBRARY_PATH=""
         export LD_LIBRARY_PATH="${KOMPILE_C_PATH}:${NATIVE_LIB_DIR}:${LD_LIBRARY_PATH}"
         if test  -f '${KOMPILE_PREFIX}./kompile-python' ; then
              rm -rf '${KOMPILE_PREFIX}./kompile-python'
         fi

         #cp -rf "${KOMPILE_PYTHON_PATH}" ./kompile-python
         cd ${KOMPILE_PREFIX}./kompile-python
         mkdir -p lib
         ${PYTHON_EXEC} setup.py build_ext --inplace
         # Work around for bundling not working properly with wheel.
         # Allow  artifacts to automatically be specified so they can be bundled.
         cp -rf ${REAL_LIB_PATH}/* ./kompile/interface/native/
          ${PYTHON_EXEC} setup.py bdist_wheel
         cd ..
         echo "Creating bundle directory ${IMAGE_NAME}-bundle"
         # Ensure old copies are removed
         if test -f "${IMAGE_NAME}-bundle" ; then
             rm -rf "${IMAGE_NAME}-bundle"
         fi
         mkdir -p "${IMAGE_NAME}-bundle"
         # Copy the include directory, library directory, python sdk, pipeline file in to the bundle
         cp -rf kompile-python/dist/*.whl "${IMAGE_NAME}-bundle"
         cp -rf "${REAL_INCLUDE_PATH}" "${IMAGE_NAME}-bundle"
         echo "Real library path is ${REAL_LIB_PATH}"
         cp -rf "${REAL_LIB_PATH}" "${IMAGE_NAME}-bundle/lib"
         if test -f "${IMAGE_NAME}-bundle/lib/${IMAGE_NAME}.${BINARY_EXTENSION}"; then
                mv "${IMAGE_NAME}-bundle/lib/${IMAGE_NAME}.${BINARY_EXTENSION}" "${IMAGE_NAME}-bundle/lib/lib${IMAGE_NAME}.${BINARY_EXTENSION}"
         fi
         cp "${PIPELINE_FILE}" "${IMAGE_NAME}-bundle"
         tar cvf "${IMAGE_NAME}-bundle.tar" "${IMAGE_NAME}-bundle"
         echo "Bundle built for image name"
      else
          echo "Built serving library. Please find executable in target/${IMAGE_NAME}"
    fi
    else
        echo "${PIPELINE_FILE} not found. Please specify a pre existing file."
        exit 1
fi
